shotover/src/codec/cassandra.rs

```diff
@@ -561,46 +659,110 @@ impl CassandraEncoder {
-                let header_start = dst.len();
+                match compression {
+                    Compression::None => {
+                        // write envelope header with dummy values for those we cant calculate till after we write the message
+                        let header_start = dst.len();
 
-                dst.extend_from_slice(&[0, 0, 0, 0, 0, 0]);
-                let payload_start = dst.len();
+                        dst.extend_from_slice(&[0, 0, 0, 0, 0, 0]);
+                        let payload_start = dst.len();
 
-                self.encode_envelope(dst, m)?;
+                        self.encode_envelope(dst, m, Compression::None)?;
 
-                //measure length of message and calculate crc24 and overwrite frame header values
-                let mut payload_len = dst.len() - payload_start;
+                        //measure length of message and calculate crc24 and overwrite frame header values
+                        let mut payload_len = (dst.len() - payload_start) as u64;
 
-                if true {
-                    // TODO if self_contained
-                    payload_len |= 1 << 17;
-                }
+                        if true {
+                            // TODO if self_contained
+                            payload_len |= 1 << 17;
+                        }
 
-                // add header length & header crc
-                let payload_len = &payload_len.to_le_bytes()[..3];
-                dst[header_start..header_start + 3].copy_from_slice(payload_len);
-                dst[header_start + 3..header_start + 6].copy_from_slice(
-                    &cassandra_protocol::crc::crc24(payload_len).to_le_bytes()[..3],
-                );
+                        // add header length & header crc
+                        let payload_len = &payload_len.to_le_bytes()[..3];
+                        dst[header_start..header_start + 3].copy_from_slice(payload_len);
+                        dst[header_start + 3..header_start + 6]
+                            .copy_from_slice(&crc24(payload_len).to_le_bytes()[..3]);
 
-                // add payload crc
-                dst.extend_from_slice(
-                    &cassandra_protocol::crc::crc32(&dst[payload_start..]).to_le_bytes(),
-                );
+                        // add payload crc
+                        dst.extend_from_slice(&crc32(&dst[payload_start..]).to_le_bytes());
+                    }
+                    Compression::Lz4 => {
+                        let header_start = dst.len();
+
+                        dst.extend_from_slice(&[0, 0, 0, 0, 0, 0, 0, 0]);
+                        let payload_start = dst.len();
+
+                        // TODO we should not be encoding small frames
+                        let (uncompressed_len, compressed_len) =
+                            self.encode_compressed_payload(dst, m, payload_start)?;
+
+                        if compressed_len > PAYLOAD_SIZE_LIMIT {
```

Comment:
we also need to check that uncompressed_len is within its limit.
It could overflow into the self contained flag

Reply:
Won't `as u64` stop that?